### ISSUE TITLE
Block cannot be moved sometimes if overlapping a custom entity.

### ISSUE THREAD

**Issue by christopho** [solarus-games/solarus#1042]
If a block is exactly on a 16x16 custom entity that does nothing and is not an obstacle, sometimes the hero cannot move the block.
The bug does not happen all the time.

This is caused by functions like Map::test_collision_with_detectors() that do not sort the results of the quadtree. The hero is facing both entities at the same time, so the chosen one is random depending on pointer values.

To avoid any performance problem and have safer code, maybe we could keep the quadtree always sorted by providing the appropriate comparator to the its sets.

**Comment by Zefk** [comment]
@christopho 
Does this have to do with NPC too? The NPC gets stuck on top of blocks when moving them if the hero is able to transverse them and the result is not being able to move the block.

```lua
elf_4:set_traversable(true)
elf_6:set_traversable(true)
```

**Example:**
![npc_movement_glitch](https://cloud.githubusercontent.com/assets/19560764/24952483/739ed1ec-1f2c-11e7-9555-175baf8cfbe0.gif)

**Comment by christopho** [comment]
No, this is unrelated. NPC cannot traverse blocks so it is normal that they are stuck. Setting the NPC traversable by the hero does not allow NPCs to traverse blocks.

**Comment by Zefk** [comment]
@christopho 
Maybe the gifs goes too fast. The NPC gets stuck on the block when the hero moves the block. Only when they are set_traversable. The block is no longer movable after that.

![npc_movement_glitch](https://cloud.githubusercontent.com/assets/19560764/24965069/d30f90e6-1f57-11e7-8cb5-78d12429eb9a.gif)

**Comment by christopho** [comment]
Ok, then please open another issue.

**Comment by Zefk** [comment]
Opened the issue. #1050

**Comment by christopho** [comment]
Same problem with other interactive entities like destructibles and chests.
